flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/SchedulingPipelinedRegionComputeUtil.java

```diff
@@ -0,0 +1,162 @@
+            for (int regionIndex : scc) {
+                mergedRegion =
+                        mergeRegions(mergedRegion, regionList.get(regionIndex), vertexToRegion);
+            }
+            mergedRegions.add(mergedRegion);
+        }
+
+        return mergedRegions;
+    }
+
+    private static List<List<Integer>> buildOutEdgesDesc(
+            final Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> vertexToRegion,
+            final List<Set<SchedulingExecutionVertex>> regionList,
+            final Function<ExecutionVertexID, ? extends SchedulingExecutionVertex>
+                    executionVertexRetriever) {
+
+        final Map<Set<SchedulingExecutionVertex>, Integer> regionIndices = new IdentityHashMap<>();
+        for (int i = 0; i < regionList.size(); i++) {
+            regionIndices.put(regionList.get(i), i);
+        }
+
+        final List<List<Integer>> outEdges = new ArrayList<>(regionList.size());
+        for (Set<SchedulingExecutionVertex> currentRegion : regionList) {
+            final List<Integer> currentRegionOutEdges = new ArrayList<>();
+            for (SchedulingExecutionVertex vertex : currentRegion) {
+                for (SchedulingResultPartition producedResult : vertex.getProducedResults()) {
+                    if (!producedResult.getResultType().isReconnectable()) {
+                        continue;
+                    }
+                    for (ConsumerVertexGroup consumerVertexGroup :
+                            producedResult.getConsumerVertexGroups()) {
+                        for (ExecutionVertexID consumerVertexId : consumerVertexGroup) {
+                            SchedulingExecutionVertex consumerVertex =
+                                    executionVertexRetriever.apply(consumerVertexId);
+                            // Skip the ConsumerVertexGroup if its vertices are outside current
+                            // regions and cannot be merged
+                            if (!vertexToRegion.containsKey(consumerVertex)) {
+                                break;
+                            }
+                            if (!currentRegion.contains(consumerVertex)) {
+                                currentRegionOutEdges.add(
+                                        regionIndices.get(vertexToRegion.get(consumerVertex)));
+                            }
+                        }
+                    }
+                }
+            }
+            outEdges.add(currentRegionOutEdges);
+        }
+
+        return outEdges;
+    }
+
+    private static Iterable<SchedulingResultPartition> getNonReconnectableConsumedResults(
+            SchedulingExecutionVertex vertex,
+            Function<IntermediateResultPartitionID, ? extends SchedulingResultPartition>
+                    resultPartitionRetriever) {
+        List<SchedulingResultPartition> nonReconnectableConsumedResults =
+                new ArrayList<>(vertex.getConsumedPartitionGroups().size());
```

Comment:
Looks to me `vertex.getConsumedPartitionGroups().size()` as the default size does not make any sense.

Reply:
Because one group can contribute 0 or multiple partitions to the result.

Reply:
Thank you for pointing this out. It seems that setting the initial size of the list the same as the count of ConsumedPartitionGroups is not good for rescale edges. I'll fix it.

Reply:
Count of `ConsumedPartitionGroups` can be either larger or smaller than the final result collection size.
In my opinion, It's incorrect and confusing.